libraries/AP_GPS/AP_GPS_SBP2.cpp

```diff
@@ -0,0 +1,494 @@
+        Debug("   ext_antenna: %d", last_heartbeat.ext_antenna);
+        Debug("   res2: %d", last_heartbeat.res2);
+        Debug("   protocol_major: %d", last_heartbeat.protocol_major);
+        Debug("   protocol_minor: %d", last_heartbeat.protocol_minor);
+        Debug("   res: %d", last_heartbeat.res);
+        Debug("   nap_error: %d", last_heartbeat.nap_error);
+        Debug("   io_error: %d", last_heartbeat.io_error);
+        Debug("   sys_error: %d", last_heartbeat.sys_error);
+
+        return false;
+
+    } else if (last_pos_llh.tow == last_vel_ned.tow
+            && abs((int32_t) (last_gps_time.tow - last_vel_ned.tow)) < 10000
+            && abs((int32_t) (last_dops.tow - last_vel_ned.tow)) < 60000
+            && last_vel_ned.tow > last_full_update_tow) {
+
+        //We have an aligned VEL and LLH, and a recent DOPS and TIME.
+
+        //
+        // Check Flags for Valid Messages
+        //
+        if (last_gps_time.flags.fix_mode    == 0 ||
+            last_vel_ned.flags.fix_mode     == 0 ||
+            last_pos_llh.flags.fix_mode     == 0 ||
+            last_dops.flags.fix_mode        == 0) {
+
+            Debug("Message Marked as Invalid. NO FIX! Flags: {GPS_TIME: %d, VEL_NED: %d, POS_LLH: %d, DOPS: %d}",
+                   last_gps_time.flags.fix_mode,
+                   last_vel_ned.flags.fix_mode,
+                   last_pos_llh.flags.fix_mode,
+                   last_dops.flags.fix_mode);
+
+            state.status = AP_GPS::NO_FIX;
+            return false;
+        }
+
+        //
+        // Update external time and accuracy state
+        //
+        state.time_week         = last_gps_time.wn;
+        state.time_week_ms      = last_vel_ned.tow;
+        state.hdop              = last_dops.hdop;
+        state.vdop              = last_dops.vdop;
+        state.last_gps_time_ms = now;
+
+        //
+        // Update velocity state
+        //
+        state.velocity[0]       = (float)(last_vel_ned.n / 1000.0);
+        state.velocity[1]       = (float)(last_vel_ned.e / 1000.0);
+        state.velocity[2]       = (float)(last_vel_ned.d / 1000.0);
+
+        float ground_vector_sq = state.velocity[0]*state.velocity[0] + state.velocity[1]*state.velocity[1];
+        state.ground_speed = safe_sqrt(ground_vector_sq);
+
+        state.ground_course = wrap_360(degrees(atan2f(state.velocity[1], state.velocity[0])));
+
+        state.speed_accuracy        = 0; //Not supported as of SBP2.1.6
+        state.horizontal_accuracy   = (float) last_pos_llh.h_accuracy*1.0e-3f;
+        state.vertical_accuracy     = (float) last_pos_llh.v_accuracy*1.0e-3f;
+
+        //
+        // Update position state
+        //
+        state.location.lat      = (int32_t) (last_pos_llh.lat * (double)1e7);
+        state.location.lng      = (int32_t) (last_pos_llh.lon * (double)1e7);
+        state.location.alt      = (int32_t) (last_pos_llh.height * 100);
```

Comment:
Do you have MSL rather then HAE? This will cause a consistent altitude offset if a Piksi device is used with a U-Blox device (or most other drivers), and result in not complying with the MAVLink spec for altitude reporting.